### PR TITLE
feat: Add withdrawal request

### PR DESCRIPTION
**Pull request by bitrocks**

### CODE REVIEW

File: packages/runner/src/runner.ts [hunk 1]
```diff
@@ -32,6 +35,10 @@ import {
 import { AlwaysSuccessGenerator } from "./locks";
 import * as secp256k1 from "secp256k1";
 
+// TODO
```
Comment (xxuejie) [inline packages/runner/src/runner.ts:38]: Please implement this TODO.

File: packages/runner/src/runner.ts [hunk 2]
```diff
@@ -475,4 +474,346 @@ export class Runner {
...
+        ckbWithdrawalCapacity + BigInt(withdrawalCapacity);
+      const sudtScriptHash = new Reader(
+        rawWithdrawalRequest.getSudtScriptHash().raw()
+      ).serializeJson();
+      let withdrawalType: Script | undefined = undefined;
+      let outputData = "0x";
+      // check if it includes sudt withdrawal request
+      if (
+        sudtScriptHash !=
+        "0x0000000000000000000000000000000000000000000000000000000000000000"
```
Comment (xxuejie) [inline packages/runner/src/runner.ts:525]: Why do we compare with a all zero script hash? I don't feel we know any script here whose hash is all zeros.

File: packages/runner/src/runner.ts [hunk 3]
```diff
@@ -475,4 +474,346 @@ export class Runner {
...
+        BigInt(outputCkbCapacitySumForSudtCustodianCells)
+    ) {
+      // so lucky with it, just return
+      return txSkeleton;
+    }
+    // build ckb change custodian cell
+    const custodianLock: Script = {
+      code_hash: this._deploymentConfig().custodian_lock.code_hash,
+      hash_type: this._deploymentConfig().custodian_lock.hash_type,
+      args: DEFAULT_CUSTODIAN_LOCK_ARGS,
```
Comment (xxuejie) [inline packages/runner/src/runner.ts:672]: If this is a change cell, can we just leverage lock args from the original custodian cell?

File: packages/runner/src/runner.ts [hunk 4]
```diff
@@ -475,4 +474,346 @@ export class Runner {
...
+          BigInt(inputCkbCapacitySum) >=
+          BigInt(ckbWithdrawalCapacity) +
+            BigInt(outputCkbCapacitySumForSudtCustodianCells) +
+            BigInt(minimalCapacity)
+        ) {
+          break;
+        }
+      }
+    }
+    // TODO collect ckb from other sudt custodian cells if all non sudt custodian cells' ckb capacity is insufficient,
```
Comment (xxuejie) [inline packages/runner/src/runner.ts:719]: I think we can add checkings in godwoken: if someone chooses to withdraw SUDT, he/she must withdraw CKB of enough capacity to store the SUDTs. Also, the left CKBs must be enough to hold a left-over custodian cell.

If above rules are not met, we should reject the withdrawal request. In this case, TODO here won't be needed.

File: packages/runner/src/runner.ts [hunk 5]
```diff
@@ -475,4 +474,346 @@ export class Runner {
...
+      BigInt(outputCkbCapacitySumForSudtCustodianCells);
+    ckbChangeCustodian.cell_output.capacity =
+      "0x" + newChangeCapacity.toString(16);
+    txSkeleton = txSkeleton.update("outputs", (outputs) => {
+      return outputs.push(ckbChangeCustodian);
+    });
+    return txSkeleton;
+  }
+
+  _buildWithdrawalLockArgs(
```
Comment (xxuejie) [inline packages/runner/src/runner.ts:741]: Please write a denormalizer for RawWithdrawalRequest, and a normalizer for WithdrawalLockArgs.

File: packages/runner/src/runner.ts (hunk 2 above)
Comment (bitrocks) [reply packages/runner/src/runner.ts:525]: In the case where user only withdraws CKB from L2, the `sudt_script_hash` would be simply set as all-zeros, and yes, it's not supposed to be found onchain. (Have confirmed it with @jjyr )

File: packages/base/src/index.ts [hunk 6]
```diff
@@ -6,11 +6,16 @@ import { CellDep, HexString, HexNumber, Hash, Script } from "@ckb-lumos/base";
...
   state_validator_type: Script;
+  sudt_type: Script;
 
   deposition_lock_dep: CellDep;
   custodian_lock_dep: CellDep;
   state_validator_lock_dep: CellDep;
   state_validator_type_dep: CellDep;
+  sudt_type_dep: CellDep;
+
+  l2_finalized_period: BigInt;
```
Comment (xxuejie) [inline packages/base/src/index.ts:20]: This is not a cell configuration, it should be in Godwoken's config, not here.

File: packages/base/src/index.ts [hunk 7]
```diff
@@ -6,11 +6,16 @@ import { CellDep, HexString, HexNumber, Hash, Script } from "@ckb-lumos/base";
 export interface DeploymentConfig {
   deposition_lock: Script;
   custodian_lock: Script;
+  withdrawal_lock: Script;
```
Comment (xxuejie) [inline packages/base/src/index.ts:9]: Add `withdrawal_lock_dep` as well.

File: packages/base/src/index.ts (hunk 6 above)
Comment (xxuejie) [reply packages/base/src/index.ts:20]: https://github.com/nervosnetwork/godwoken/blob/a0d93bc8af5ecadfd9491cdf859803b72e120c0c/crates/common/src/lib.rs#L18 We can set it as a hardcoded value like the Rust side for now, then later refactor it to a config variable.

Comment (bitrocks) [reply packages/base/src/index.ts]: updated.

File: packages/base/src/index.ts (hunk 7 above)
Comment (bitrocks) [reply packages/base/src/index.ts:17]: updated.

File: packages/runner/src/runner.ts (hunk 1 above)
Comment (bitrocks) [reply packages/runner/src/runner.ts]: updated.

File: packages/runner/src/runner.ts (hunk 3 above)
Comment (bitrocks) [reply packages/runner/src/runner.ts]: After future discussion, we agreed with the default value approach.

File: packages/runner/src/runner.ts (hunk 4 above)
Comment (bitrocks) [reply packages/runner/src/runner.ts]: updated.

File: packages/runner/src/runner.ts (hunk 5 above)
Comment (bitrocks) [reply packages/runner/src/runner.ts:851]: updated.

File: crates/common/src/lib.rs [hunk 8]
```diff
@@ -27,6 +27,7 @@ pub const CKB_SUDT_SCRIPT_HASH: [u8; 32] = [
 pub const CKB_SUDT_SCRIPT_ARGS: [u8; 32] = [0; 32];
 pub const ACCOUNT_LOCK_CODE_HASH: [u8; 32] = [0u8; 32];
 pub const ROLLUP_LOCK_CODE_HASH: [u8; 32] = [0u8; 32];
+pub const CKB_NON_SUDT_CHECK_ARGS: [u8; 32] = [0u8; 32];
```
Comment (jjyr) [inline crates/common/src/lib.rs:30]: `CKB_SUDT_SCRIPT_ARGS` should be enough to determine a SUDT and CKB.

File: crates/generator/src/generator.rs [hunk 9]
```diff
@@ -114,14 +116,26 @@ impl Generator {
...
-        if amount > balance {
+        // check CKB balance
+        let ckb_balance = state.get_sudt_balance(CKB_SUDT_ACCOUNT_ID, id)?;
+        if capacity > ckb_balance as u64 {
             return Err(ValidateError::InvalidWithdrawal.into());
         }
+        // check sudt balance
+        // when sudt_script_hash is [0u8;32] means only withdraw ckb in the withdrawal_request,
+        // otherwise check the sudt existence and balance.
+        if sudt_script_hash != CKB_NON_SUDT_CHECK_ARGS {
```
Comment (jjyr) [inline crates/generator/src/generator.rs:127]: Use CKB_SUDT_SCRIPT_ARGS

**Review by jjyr — CHANGES_REQUESTED** [review]

File: crates/generator/src/error.rs [hunk 10]
```diff
@@ -62,6 +62,8 @@ pub enum WithdrawalError {
     InvalidNonce { expected: u32, actual: u32 },
     #[error("Withdrawal Faked CKB")]
     WithdrawFakedCKB,
+    #[error("Non positive sudt amount")]
+    NonPositiveSudtAmount,
```
Comment (jjyr) [inline crates/generator/src/error.rs:66]: Use `WithdrawalError::Overdraft`.

File: crates/generator/src/generator.rs [hunk 11]
```diff
@@ -129,6 +129,9 @@ impl Generator {
             .ok_or(ValidateError::UnknownSUDT)?;
         if sudt_id != CKB_SUDT_ACCOUNT_ID {
             // check SUDT balance
+            if amount <= 0 {
```
Comment (jjyr) [inline crates/generator/src/generator.rs:132]: Remove this, the code below already does the verification.

**Review by jjyr — CHANGES_REQUESTED** [review]

File: crates/chain/src/chain.rs [hunk 12]
```diff
@@ -629,6 +629,10 @@ impl Chain {
         };
         // reverted_block_root: Byte32,
         let last_finalized_block_number = number.saturating_sub(FINALIZE_BLOCKS);
+        println!(
```
[the remark this answers is not in the document]
Comment (bitrocks) [reply crates/chain/src/chain.rs]: updated.

File: crates/generator/src/generator.rs (hunk 11 above)
Comment (bitrocks) [reply crates/generator/src/generator.rs]: User can't withdrawal 0 SUDT when non-CKB sudt_id exists.

File: crates/generator/src/error.rs (hunk 10 above)
Comment (jjyr) [inline crates/generator/src/error.rs:66]: rename Sudt -> SUDT to keep consistency

Comment (bitrocks) [reply crates/generator/src/error.rs]: updated.

**Review by jjyr — APPROVED** [review]

**Pull request merged**